src/provider/model/AppWindow.ts

```diff
@@ -53,5 +64,192 @@ export interface AppWindow {
+        this._intentListeners.add(intentName);
+        this._onIntentListenerAdded.emit(intentName);
+    }
+
+    public removeIntentListener(intentName: string): void {
+        this._intentListeners.delete(intentName);
+    }
+
+    public hasContextListener(): boolean {
+        return this._contextListener;
+    }
+
+    public addContextListener(): void {
+        this._contextListener = true;
+        this._onContextListenerAdded.emit();
+    }
+
+    public removeContextListener(): void {
+        this._contextListener = false;
+    }
+
+    public hasChannelContextListener(channel: ContextChannel): boolean {
+        return this._channelContextListeners.has(channel.id);
+    }
+
+    public addChannelContextListener(channel: ContextChannel): void {
+        this._channelContextListeners.add(channel.id);
+    }
+
+    public removeChannelContextListener(channel: ContextChannel): void {
+        this._channelContextListeners.delete(channel.id);
+    }
+
+    public hasChannelEventListener(channel: ContextChannel, eventType: ChannelEvents['type']): boolean {
+        return this._channelEventListeners.has(channel.id) && (this._channelEventListeners.get(channel.id)!.has(eventType));
+    }
+
+    public addChannelEventListener(channel: ContextChannel, eventType: ChannelEvents['type']): void {
+        if (!this._channelEventListeners.has(channel.id)) {
+            this._channelEventListeners.set(channel.id, new Set());
+        }
+
+        this._channelEventListeners.get(channel.id)!.add(eventType);
+    }
+
+    public removeChannelEventListener(channel: ContextChannel, eventType: ChannelEvents['type']): void {
+        if (this._channelEventListeners.has(channel.id)) {
+            const events = this._channelEventListeners.get(channel.id)!;
+            events.delete(eventType);
+        }
+    }
+
+    public async isReadyToReceiveIntent(intent: IntentType): Promise<boolean> {
+        if (this.hasIntentListener(intent)) {
+            // App has already registered the intent listener
+            return true;
+        }
+
+        const age = this._creationTime === undefined ? undefined : Date.now() - this._creationTime;
+
+        if (age === undefined || age >= Timeouts.ADD_INTENT_LISTENER) {
+            // App has been running for a while
+            return false;
+        } else {
+            // App may be starting - Give it some time to initialize and call `addIntentListener()`, otherwise timeout
+            const deferredPromise = new DeferredPromise();
+
+            const slot = this._onIntentListenerAdded.add(intentAdded => {
+                if (intentAdded === intent) {
+                    deferredPromise.resolve();
+                }
+            });
+
+            const [didTimeout] = await withTimeout(Timeouts.ADD_INTENT_LISTENER - age, deferredPromise.promise);
+
+            slot.remove();
+
+            return !didTimeout;
+        }
+    }
+
+    public async isReadyToReceiveContext(): Promise<boolean> {
```

Comment:
Can shared code be pulled out and used by both `isReadyToReceiveIntent` and `isReadyToReceiveContext`? or does it wind up being too messy to be worthwhile?

Reply:
Hmmm, maybe. My concern is that there are subtle differences pretty much throughout these functions, which would make whatever common function have several `if context` / `if intent` block, at which point I think we lose whatever value we gain from consolidating. Specifically, we would need to switch for the `has[x]Listener` function, the timeout constant, and the `on[x]ListenerAdded` signal, which to me pushes it into the "too messy" region.